{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "831a5617",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use conda env open2c_env.yml to create conda env for this script\n",
    "#This is an example script of plotting MC-3C data\n",
    "#It is run from within the 'scripts' subdirectory, using following directory structure:\n",
    "#Analysis_Dir\n",
    "#├── data\n",
    "#    ├── permutations\n",
    "#├── alignments\n",
    "#├── figures\n",
    "#├── scripts\n",
    "#├── lsf_jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f07ee29",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioframe\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib import cm\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib.gridspec import GridSpecFromSubplotSpec\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib.colors import ListedColormap\n",
    "import random\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "import pickle\n",
    "from numpy import diff\n",
    "\n",
    "from pandas import read_csv\n",
    "from sklearn.utils import resample\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.metrics import accuracy_score\n",
    "from matplotlib import pyplot\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f22625",
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = [\n",
    "    't0Mit_R1',\n",
    "    't2_R1',\n",
    "    't4DMSO_R1',\n",
    "    't4ICRF_R1',\n",
    "    't8DMSO_R1',\n",
    "    't8ICRF_R1',\n",
    "    't0Mit_R2',\n",
    "    't2_R2',\n",
    "    't4DMSO_R2',\n",
    "    't4ICRF_R2',\n",
    "    't8DMSO_R2',\n",
    "    't8ICRF_R2',\n",
    "    't0Mit_R3',\n",
    "    't2_R3',\n",
    "    't4DMSO_R3',\n",
    "    't4ICRF_R3',\n",
    "    't8DMSO_R3',\n",
    "    't8ICRF_R3'\n",
    "]\n",
    "\n",
    "long_names = {\n",
    "    't0Mit_R1' : 'TI-MC3C-Dpn-t0Mit-4-30',\n",
    "    't2_R1' : 'TI-MC3C-Dpn-t2-4-30',\n",
    "    't4DMSO_R1' : 'TI-MC3C-Dpn-t4DMSO-4-30',\n",
    "    't4ICRF_R1' : 'TI-MC3C-Dpn-t4ICRF-4-30',\n",
    "    't8DMSO_R1' : 'TI-MC3C-Dpn-t8DMSO-4-30',\n",
    "    't8ICRF_R1' : 'TI-MC3C-Dpn-t8ICRF-4-30',\n",
    "    't0Mit_R2' : 'TI-MC3C-Dpn-t0Mit-4-39',\n",
    "    't2_R2' : 'TI-MC3C-Dpn-t2-4-39',\n",
    "    't4DMSO_R2' : 'TI-MC3C-Dpn-t4DMSO-4-39',\n",
    "    't4ICRF_R2' : 'TI-MC3C-Dpn-t4ICRF-4-39',\n",
    "    't8DMSO_R2' : 'TI-MC3C-Dpn-t8DMSO-4-39',\n",
    "    't8ICRF_R2' : 'TI-MC3C-Dpn-t8ICRF-4-39',\n",
    "    't0Mit_R3' : 'TI-MC3C-Dpn-t0Mit-R3-5-14',\n",
    "    't2_R3' : 'TI-MC3C-Dpn-t2-R3-5-14',\n",
    "    't4DMSO_R3' : 'TI-MC3C-Dpn-t4DMSO-R3-5-14',\n",
    "    't4ICRF_R3' : 'TI-MC3C-Dpn-t4ICRF-R3-5-14',\n",
    "    't8DMSO_R3' : 'TI-MC3C-Dpn-t8DMSO-R3-5-14',\n",
    "    't8ICRF_R3' : 'TI-MC3C-Dpn-t8ICRF-R3-5-14',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa973b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleColors = {\n",
    "    't0Mit_R1' : '#878787',\n",
    "    't2_R1' : '#E1B7A3',\n",
    "    't4DMSO_R1' : '#17BECF',\n",
    "    't4ICRF_R1' : '#D62728',\n",
    "    't8DMSO_R1': '#0D6871',\n",
    "    't8ICRF_R1': '#751616',\n",
    "    't0Mit_R2' : '#878787',\n",
    "    't2_R2' : '#E1B7A3',\n",
    "    't4DMSO_R2' : '#17BECF',\n",
    "    't4ICRF_R2' : '#D62728',\n",
    "    't8DMSO_R2' :  '#0D6871',\n",
    "    't8ICRF_R2' : '#751616',\n",
    "    't0Mit_R3' : '#878787',\n",
    "    't2_R3' : '#E1B7A3',\n",
    "    't4DMSO_R3' : '#17BECF',\n",
    "    't4ICRF_R3' : '#D62728',\n",
    "    't8DMSO_R3' :  '#0D6871',\n",
    "    't8ICRF_R3' : '#751616'\n",
    "}\n",
    "    \n",
    "sampleLineStyles = {\n",
    "    't0Mit_R1' : '-',\n",
    "    't2_R1' : '-',\n",
    "    't4DMSO_R1' : '-',\n",
    "    't4ICRF_R1' : '-',\n",
    "    't8DMSO_R1': '-',\n",
    "    't8ICRF_R1': '-',\n",
    "    't0Mit_R2' : '--',\n",
    "    't2_R2' : '--',\n",
    "    't4DMSO_R2' : '--',\n",
    "    't4ICRF_R2' : '--',\n",
    "    't8DMSO_R2' :  '--',\n",
    "    't8ICRF_R2' : '--',\n",
    "    't0Mit_R3' : ':',\n",
    "    't2_R3' : ':',\n",
    "    't4DMSO_R3' : ':',\n",
    "    't4ICRF_R3' : ':',\n",
    "    't8DMSO_R3' : ':',\n",
    "    't8ICRF_R3' : ':',\n",
    "}\n",
    "\n",
    "samplePlotNames = {\n",
    "    't0Mit_R1' : 't0 Mit',\n",
    "    't2_R1' : 't2',\n",
    "    't4DMSO_R1' : 't4 DMSO',\n",
    "    't4ICRF_R1' : 't4 ICRF-193',\n",
    "    't8DMSO_R1' : 't8 DMSO',\n",
    "    't8ICRF_R1' : 't8 ICRF-193',\n",
    "    't0Mit_R2' : 't0 Mit, R2',\n",
    "    't2_R2' : 't2, R2',\n",
    "    't4DMSO_R2' : 't4 DMSO, R2',\n",
    "    't4ICRF_R2' : 't4 ICRF-193, R2',\n",
    "    't8DMSO_R2' : 't8 DMSO, R2',\n",
    "    't8ICRF_R2' : 't8 ICRF-193, R2',\n",
    "    't0Mit_R3' : 't0 Mit, R3',\n",
    "    't2_R3' : 't2, R3',\n",
    "    't4DMSO_R3' : 't4 DMSO, R3',\n",
    "    't4ICRF_R3' : 't4 ICRF-193, R3',\n",
    "    't8DMSO_R3' : 't8 DMSO, R3',\n",
    "    't8ICRF_R3' : 't8 ICRF-193, R3' \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6e45317",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrlconds = [\n",
    "    't8DMSO_R1',\n",
    "    't8DMSO_R1',\n",
    "    't8DMSO_R1',\n",
    "    't8DMSO_R1',\n",
    "    't8DMSO_R1',\n",
    "    't4DMSO_R1',\n",
    "    't8DMSO_R2',\n",
    "    't8DMSO_R2',\n",
    "    't8DMSO_R2',\n",
    "    't8DMSO_R2',\n",
    "    't8DMSO_R2',\n",
    "    't4DMSO_R2',\n",
    "    't8DMSO_R3',\n",
    "    't8DMSO_R3',\n",
    "    't8DMSO_R3',\n",
    "    't8DMSO_R3',\n",
    "    't8DMSO_R2',\n",
    "    't4DMSO_R3',\n",
    "]\n",
    "\n",
    "treatconds = [\n",
    "    't0Mit_R1',\n",
    "    't2_R1',\n",
    "    't4DMSO_R1',\n",
    "    't4ICRF_R1',\n",
    "    't8ICRF_R1',\n",
    "    't4ICRF_R1',\n",
    "    't0Mit_R2',\n",
    "    't2_R2',\n",
    "    't4DMSO_R2',\n",
    "    't4ICRF_R2',\n",
    "    't8ICRF_R2',\n",
    "    't4ICRF_R2',\n",
    "    't0Mit_R3',\n",
    "    't2_R3',\n",
    "    't4DMSO_R3',\n",
    "    't4ICRF_R3',\n",
    "    't8ICRF_R3',\n",
    "    't4ICRF_R3',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b09ff163",
   "metadata": {},
   "outputs": [],
   "source": [
    "repdict = {\n",
    "    't0Mit_R1' : 'R1',\n",
    "    't2_R1' : 'R1',\n",
    "    't4DMSO_R1' : 'R1',\n",
    "    't4ICRF_R1' : 'R1',\n",
    "    't8DMSO_R1' : 'R1',\n",
    "    't8ICRF_R1' : 'R1',\n",
    "    't0Mit_R2' : 'R2',\n",
    "    't2_R2' : 'R2',\n",
    "    't4DMSO_R2' : 'R2',\n",
    "    't4ICRF_R2' : 'R2',\n",
    "    't8DMSO_R2' : 'R2',\n",
    "    't8ICRF_R2' : 'R2',\n",
    "    't0Mit_R3' : 'R3',\n",
    "    't2_R3' : 'R3',\n",
    "    't4DMSO_R3' : 'R3',\n",
    "    't4ICRF_R3' : 'R3',\n",
    "    't8DMSO_R3' : 'R3',\n",
    "    't8ICRF_R3' : 'R3',\n",
    "}\n",
    "\n",
    "labeldict = {\n",
    "    't0Mit_R1' : 't0 Prometa',\n",
    "    't2_R1' : 't2 Ana/Telo',\n",
    "    't4DMSO_R1' : 't4 G1 DMSO',\n",
    "    't4ICRF_R1' : 't4 G1 ICRF-193',\n",
    "    't8DMSO_R1' : 't8 G1 DMSO',\n",
    "    't8ICRF_R1' : 't8 G1 ICRF-193',\n",
    "    't0Mit_R2' : 't0 Prometa',\n",
    "    't2_R2' : 't2 Ana/Telo',\n",
    "    't4DMSO_R2' : 't4 G1 DMSO',\n",
    "    't4ICRF_R2' : 't4 G1 ICRF-193',\n",
    "    't8DMSO_R2' : 't8 G1 DMSO',\n",
    "    't8ICRF_R2' : 't8 G1 ICRF-193',\n",
    "    't0Mit_R3' : 't0 Prometa',\n",
    "    't2_R3' : 't2 Ana/Telo',\n",
    "    't4DMSO_R3' : 't4 G1 DMSO',\n",
    "    't4ICRF_R3' : 't4 G1 ICRF-193',\n",
    "    't8DMSO_R3' : 't8 G1 DMSO',\n",
    "    't8ICRF_R3' : 't8 G1 ICRF-193',\n",
    "}\n",
    "\n",
    "labelPlotColors = {\n",
    "    't0 Prometa' : '#878787',\n",
    "    't2 Ana/Telo' : '#E1B7A3',\n",
    "    't4 G1 DMSO' : '#17BECF',\n",
    "    't4 G1 ICRF-193' : '#D62728',\n",
    "    't8 G1 DMSO': '#0D6871',\n",
    "    't8 G1 ICRF-193': '#751616',\n",
    "}\n",
    "\n",
    "labels = [\n",
    "    't0 Prometa',\n",
    "    't2 Ana/Telo',\n",
    "    't4 G1 DMSO',\n",
    "    't4 G1 ICRF-193',\n",
    "    't8 G1 DMSO',\n",
    "    't8 G1 ICRF-193',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa498c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For each of the above, can separate by A vs B compartments, or by number of chromosomes visited\n",
    "comp_types = ['A', 'B', 'AB']\n",
    "good_chroms = ['chr4', 'chr14', 'chr17', 'chr18', 'chr20', 'chr21']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4a68a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "outDataDir = '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c297db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in pickled files\n",
    "\n",
    "f = open(f'{outDataDir}/data/220517_MRICRF_R1R2R3_real_walks_with_summary_firstx_length_fractions_dict_100kbEigs.pkl', 'rb')\n",
    "real_walks_with_summary_firstx = pickle.load(f)\n",
    "f.close()\n",
    "\n",
    "f = open(f'{outDataDir}/data/220517_MRICRF_R1R2R3_real_walks_filtered_summarized_firstx_length_dict_100kbEigs.pkl', 'rb')\n",
    "real_walks_summarized_firstx = pickle.load(f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b3e918",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combining T1 and T2 reads for t0 Mit R1\n",
    "real_walks_with_summary_firstx['t0Mit_R1'] = {}\n",
    "for fragnum in [2, 3, 4, 5, 6, 7, 8, 9, 10]:\n",
    "    real_walks_with_summary_firstx['t0Mit_R1'][f'length_{fragnum}'] = real_walks_with_summary_firstx['t0Mit_R1_T1'][f'length_{fragnum}'].append(\n",
    "        real_walks_with_summary_firstx['t0Mit_R1_T2'][f'length_{fragnum}'],\n",
    "        ignore_index = True\n",
    "    ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edcfe991",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combining T1 and T2 reads for t0 Mit R1\n",
    "real_walks_summarized_firstx['t0Mit_R1'] = {}\n",
    "for fragnum in [2, 3, 4, 5, 6, 7, 8, 9, 10]:\n",
    "    real_walks_summarized_firstx['t0Mit_R1'][f'length_{fragnum}'] = real_walks_summarized_firstx['t0Mit_R1_T1'][f'length_{fragnum}'].append(\n",
    "        real_walks_summarized_firstx['t0Mit_R1_T2'][f'length_{fragnum}'],\n",
    "        ignore_index = True\n",
    "    ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7538a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Kernel density plots of pairwise distance - overlay - 1 or 2 chrom walks, A, B, or AB compartments\n",
    "#Just good chromosome walks here\n",
    "#Just first 6 steps of walks\n",
    "\n",
    "dist_df = pd.DataFrame()\n",
    "\n",
    "#combine biological replicates\n",
    "for i, cond in enumerate(conditions):\n",
    "    df = real_walks_with_summary_firstx[cond]['length_6']\n",
    "    df['Query_Fragment_Length'] = df['Query_End'] - df['Query_Start']\n",
    "    grouped_walks = df.groupby('Query_Name')\n",
    "    walks_min_mapq = grouped_walks.agg({'Mapping_Quality' : 'min'})\n",
    "    good_walks_mapq = walks_min_mapq[walks_min_mapq['Mapping_Quality'] > 59] #use this to filter for mapq\n",
    "    walks_frac_map = grouped_walks.agg({'Match_Length' : 'sum',\n",
    "                                    'Query_Fragment_Length' : 'sum',\n",
    "                                    'Alignment_Length' : 'sum'\n",
    "                                   })\n",
    "    walks_high_frac_map = walks_frac_map[\n",
    "        (walks_frac_map['Match_Length']/walks_frac_map['Query_Fragment_Length']) > 0.8]\n",
    "    #use this to filter for fraction mapped\n",
    "\n",
    "    cond1_filtered = df[\n",
    "        (df['Chrom_Number'] < 3) &\n",
    "        (df['Walk_Comp_Type'].isin(['A', 'B', 'AB'])) &\n",
    "        (df['Query_Name'].isin(good_walks_mapq.index)) &\n",
    "        (df['Query_Name'].isin(walks_high_frac_map.index)) &\n",
    "        (df['chrom'].isin(good_chroms))]\n",
    "    dist_dropna = pd.DataFrame({'Dist' : cond1_filtered['dist'].copy(),\n",
    "                               'Condition' : cond,\n",
    "                               'Label' : labeldict[cond],\n",
    "                               'Replicate' : repdict[cond]})\n",
    "    dist_dropna = dist_dropna.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "    dist_dropna['Abs_Dist'] = abs(dist_dropna['Dist']).replace([np.inf, -np.inf], np.nan).dropna()    \n",
    "    dist_df = dist_df.append(dist_dropna[dist_dropna['Abs_Dist'] > 0], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05768f16",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df, x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')\n",
    "plt.savefig(f'{outDataDir}/figures/220518_AllSamples_R1R2R3_Combined_Density_Dist_LogScaleKde_DropNA.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f174872a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df, x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('MC-3C Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "plt.xlim(1e2, 1e9)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')\n",
    "plt.savefig(f'{outDataDir}/figures/220518_AllSamples_R1R2R3_Combined_Density_Dist_LogScaleKde_DropNA_MatchHiC_Axes.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ce60f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f50665",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculating location of peaks in each dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3288340a",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_list = pd.DataFrame(columns = ['Label', 'PeakNum', 'PeakSize(bp)', 'Prominence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "410c4b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't0 Prometa'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eccdf041",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't0 Prometa',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594dcaf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't2 Ana/Telo'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b69dc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't2 Ana/Telo',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af5fab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't4 G1 DMSO'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "508fa603",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't4 G1 DMSO',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f602b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't4 G1 ICRF-193'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1907e594",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't4 G1 ICRF-193',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4c0e276",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't8 G1 DMSO'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c5d9072",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't8 G1 DMSO',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d1dc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = dist_df[dist_df['Label'] == 't8 G1 ICRF-193'], x = 'Abs_Dist', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Pairwise Interaction Distance')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b809485f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.001)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't8 G1 ICRF-193',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "becbd921",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea346052",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    g = sns.displot(kind='kde', data=dist_df, col='Replicate', x='Abs_Dist', hue='Label', \n",
    "                    palette = cmap_bar, log_scale = True, common_norm = False, common_grid = True, \n",
    "                    lw = 2, facet_kws={'sharey': True}, height = 3)\n",
    "    g._legend.set_title('Sample')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_AllSamples_R1R2R3_Separate_Density_Dist_LogScaleKde_DropNA.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cce6cafc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#walk span plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "494cfe97",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Kernel density plots of walk span - overlay - 1 or 2 chrom walks, A, B, or AB compartments\n",
    "#Just good chromosome walks here\n",
    "#Just first 6 steps of walks\n",
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "\n",
    "span_df = pd.DataFrame()\n",
    "\n",
    "#combine biological replicates\n",
    "for i, cond in enumerate(conditions):\n",
    "    df = real_walks_with_summary_firstx[cond]['length_6']\n",
    "    df['Query_Fragment_Length'] = df['Query_End'] - df['Query_Start']\n",
    "    grouped_walks = df.groupby('Query_Name')\n",
    "    walks_min_mapq = grouped_walks.agg({'Mapping_Quality' : 'min'})\n",
    "    good_walks_mapq = walks_min_mapq[walks_min_mapq['Mapping_Quality'] > 59] #use this to filter for mapq\n",
    "    walks_frac_map = grouped_walks.agg({'Match_Length' : 'sum',\n",
    "                                    'Query_Fragment_Length' : 'sum',\n",
    "                                    'Alignment_Length' : 'sum'})\n",
    "    walks_high_frac_map = walks_frac_map[\n",
    "        (walks_frac_map['Match_Length']/walks_frac_map['Query_Fragment_Length']) > 0.8]\n",
    "    #use this to filter for fraction mapped\n",
    "\n",
    "    df2 = real_walks_summarized_firstx[cond]['length_6']\n",
    "    cond1_filtered = df2[\n",
    "        (df2['Chrom_Number'] == 1) &\n",
    "        (df2['Walk_Comp_Type'].isin(['A', 'B', 'AB'])) &\n",
    "        (df2['Query_Name'].isin(good_walks_mapq.index)) &\n",
    "        (df2['Query_Name'].isin(walks_high_frac_map.index)) &\n",
    "        (df2['chrom'].isin(good_chroms))]\n",
    "    dist_dropna = pd.DataFrame({'Span' : cond1_filtered['Span'].copy(),\n",
    "                               'Condition' : cond,\n",
    "                               'Label' : labeldict[cond],\n",
    "                               'Replicate' : repdict[cond]})\n",
    "    dist_dropna = dist_dropna.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "    dist_dropna['Abs_Span'] = abs(dist_dropna['Span']).replace([np.inf, -np.inf], np.nan).dropna()     \n",
    "    span_df = span_df.append(dist_dropna[dist_dropna['Abs_Span'] > 0], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e87ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Kernel density plots - overlay - 1 chrom walks, A, B, or AB compartments\n",
    "fig = plt.figure(figsize=(3, 3))\n",
    "\n",
    "g = sns.kdeplot(data = span_df, x = 'Abs_Span', common_norm = False,\n",
    "        hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')\n",
    "plt.xlabel('Walk Span, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/220518_MRICRF_AllSamples_R1R2R3_Density_Span_Combo_1st6steps_Logbins_DropNA_100kbEigs_GoodChroms.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35a85bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#peak list for walk span as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0060b8c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_list = pd.DataFrame(columns = ['Label', 'PeakNum', 'PeakSize(bp)', 'Prominence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc26c3d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't0 Prometa'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "913a82cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't0 Prometa',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b61f793c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't2 Ana/Telo'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1099d744",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't2 Ana/Telo',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bb8f245",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't4 G1 DMSO'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d071ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't4 G1 DMSO',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4a15bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't4 G1 ICRF-193'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac77639d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't4 G1 ICRF-193',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099d07f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't8 G1 DMSO'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "856da5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't8 G1 DMSO',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3ce4c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3, 3))\n",
    "cmap_bar = sns.color_palette(['#878787'])\n",
    "\n",
    "g = sns.kdeplot(data = span_df[span_df['Label'] == 't8 G1 ICRF-193'], x = 'Abs_Span', common_norm = False,\n",
    "    hue = 'Label', lw = 2, log_scale = True, common_grid = True, palette = cmap_bar)\n",
    "plt.xlabel('Separation, bp')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Walk Span')\n",
    "plt.grid(lw=0.5)\n",
    "sns.move_legend(g, \"upper left\", bbox_to_anchor=(1.05, 1), title='Sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9f10eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Location of peaks\n",
    "#from https://stackoverflow.com/questions/63492366/python-get-fwhm-from-seaborn-kdeplot\n",
    "#https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html\n",
    "\n",
    "kde_curve = g.lines[0]\n",
    "x = kde_curve.get_xdata()\n",
    "y = kde_curve.get_ydata()\n",
    "peaks, properties = scipy.signal.find_peaks(y, prominence = 0.01)\n",
    "for i, element in enumerate(peaks):\n",
    "    xpos = x[element]\n",
    "    prom = properties['prominences'][i]\n",
    "    peak_list = peak_list.append({\n",
    "        'Label' : 't8 G1 ICRF-193',\n",
    "        'PeakNum' : i,\n",
    "        'PeakSize(bp)' : xpos,\n",
    "        'Prominence' : prom\n",
    "    }, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80bd785f",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d105d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    g = sns.displot(kind='kde', data=span_df, col='Replicate', x='Abs_Span', hue='Label', \n",
    "                    palette = cmap_bar, log_scale = True, common_norm = False, common_grid = True, \n",
    "                    lw = 2, facet_kws={'sharey': True}, height = 3)\n",
    "    g._legend.set_title('Sample')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_AllSamples_R1R2R3_Separate_Density_Span_LogScaleKde_DropNA.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c85963b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fraction of 1 vs >1 chromosome walks in A, B, AB\n",
    "\n",
    "CompartmentType_ByInterChromSteps = pd.DataFrame()\n",
    "\n",
    "for cond in conditions:\n",
    "    df = real_walks_with_summary_firstx[cond]['length_6']\n",
    "    df['Query_Fragment_Length'] = df['Query_End'] - df['Query_Start']\n",
    "    grouped_walks = df.groupby('Query_Name')\n",
    "    walks_min_mapq = grouped_walks.agg({'Mapping_Quality' : 'min'})\n",
    "    good_walks_mapq = walks_min_mapq[walks_min_mapq['Mapping_Quality'] > 59] #use this to filter for mapq\n",
    "    walks_frac_map = grouped_walks.agg({'Match_Length' : 'sum',\n",
    "                                    'Query_Fragment_Length' : 'sum',\n",
    "                                    'Alignment_Length' : 'sum'\n",
    "                                   })\n",
    "    walks_high_frac_map = walks_frac_map[\n",
    "        (walks_frac_map['Match_Length']/walks_frac_map['Query_Fragment_Length']) > 0.8] #use this to filter for fraction mapped\n",
    "\n",
    "    df2 = real_walks_summarized_firstx[cond]['length_6']\n",
    "    cond1_filtered = df2[\n",
    "        (df2['Chrom_Number'] < 3) &\n",
    "        (df2['Walk_Comp_Type'].isin(['A', 'B', 'AB'])) &\n",
    "        (df2['Query_Name'].isin(good_walks_mapq.index)) &\n",
    "        (df2['Query_Name'].isin(walks_high_frac_map.index)) &\n",
    "        (df2['chrom'].isin(good_chroms))]\n",
    "    \n",
    "    #1 chrom walks\n",
    "    cis_walks = cond1_filtered[\n",
    "        (cond1_filtered['Chrom_Number'] == 1)]\n",
    "    cis_walks['Condition'] = cond\n",
    "    cis_walks['Label'] = labeldict[cond]\n",
    "    cis_walks['Replicate'] = repdict[cond]\n",
    "    cis_walks['Walk_CisTrans_Type'] = '1 chrom'\n",
    "    \n",
    "    #2 chrom walks\n",
    "    trans_walks = cond1_filtered[\n",
    "        (cond1_filtered['Chrom_Number'] == 2)]\n",
    "    trans_walks['Condition'] = cond\n",
    "    trans_walks['Label'] = labeldict[cond]\n",
    "    trans_walks['Replicate'] = repdict[cond]\n",
    "    trans_walks['Walk_CisTrans_Type'] = '2 chrom'\n",
    "    \n",
    "    CompartmentType_ByInterChromSteps = CompartmentType_ByInterChromSteps.append(cis_walks, ignore_index = True)\n",
    "    CompartmentType_ByInterChromSteps = CompartmentType_ByInterChromSteps.append(trans_walks, ignore_index = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31de6025",
   "metadata": {},
   "outputs": [],
   "source": [
    "CompartmentType_ByInterChromSteps_ct = pd.crosstab(\n",
    "    index=[CompartmentType_ByInterChromSteps['Condition']],\n",
    "    columns=[CompartmentType_ByInterChromSteps['Chrom_Number'],\n",
    "             CompartmentType_ByInterChromSteps['Walk_Comp_Type']],\n",
    "    normalize = 'index')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576cf696",
   "metadata": {},
   "outputs": [],
   "source": [
    "CompartmentType_ByInterChromSteps_ct.plot(kind = 'bar', stacked = True, use_index = True,\n",
    "                                          colormap='Accent', figsize = (7, 4))\n",
    "plt.xlabel('Sample')\n",
    "plt.ylabel('Fraction of Walks') \n",
    "plt.title('Walk Types')\n",
    "plt.legend(bbox_to_anchor=(1.04,1), title = '# of Chroms, Comp. Type')\n",
    "     \n",
    "plt.savefig(f'{outDataDir}/figures/220518_MRICRF_AllSamples_R1R2R3_MC3C_CisTransByCompType_Bar_100kbbinEig1_GoodChroms.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cfc4a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "CompartmentType_ByInterChromSteps_ct = pd.crosstab(\n",
    "    index=[CompartmentType_ByInterChromSteps['Condition']\n",
    "          ],\n",
    "    columns=[CompartmentType_ByInterChromSteps['Chrom_Number'],\n",
    "             CompartmentType_ByInterChromSteps['Walk_Comp_Type']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d03367",
   "metadata": {},
   "outputs": [],
   "source": [
    "CompartmentType_ByInterChromSteps_ct.sum(axis = 1) #good chroms only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ca420ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scaling plots and scaling slope"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b9781e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bedslice(grouped, chrom, start, end):\n",
    "\t\"\"\"Assumes no proper nesting of intervals\"\"\"\n",
    "\tchromdf = grouped.get_group(chrom)\n",
    "\tlo = chromdf['end'].values.searchsorted(start, side='right')\n",
    "\thi = lo + chromdf['start'].values[lo:].searchsorted(end, side='left')\n",
    "\treturn chromdf.iloc[lo:hi]\n",
    "\n",
    "\n",
    "def intlogbins(start, end, N=None, ratio=None):\n",
    "\t\n",
    "\tstart = int(start)\n",
    "\tend = int(end)\n",
    "\tif ratio is not None:\n",
    "\t\tif N is not None:\n",
    "\t\t\traise ValueError(\"Please specify only N or ratio\")\n",
    "\t\tN = np.log(end / start) / np.log(ratio)\n",
    "\telif N is None:\n",
    "\t\traise ValueError(\"Please specify either N or ratio\")\n",
    "\tbins = np.logspace(np.log10(start), np.log10(end), N)\n",
    "\tbins = np.array(np.rint(bins), dtype=int)\n",
    "\tfor i in range(1,len(bins)-1):\n",
    "\t\tif bins[i]==bins[i-1]:\n",
    "\t\t\tbins[i] = bins[i]+1\n",
    "\tassert np.all(bins[1:] > bins[:-1])\n",
    "\tassert bins[0] == start\n",
    "\tassert bins[-1] == end\n",
    "\treturn bins\n",
    "\n",
    "\n",
    "def geomprog(factor, start=1):\n",
    "\tyield start\n",
    "\twhile True:\n",
    "\t\tstart *= factor\n",
    "\t\tyield start\n",
    "\n",
    "\n",
    "def _geomrange(start, end, factor, endpoint):\n",
    "\tprev = np.nan\n",
    "\tfor i in geomprog(factor, start):\n",
    "\t\tx = int(round(i))\n",
    "\t\t\n",
    "\t\tif x > end:\n",
    "\t\t\tbreak\n",
    "\n",
    "\t\tif x == prev:\n",
    "\t\t\tcontinue\n",
    "\t\t\n",
    "\t\tprev = x\n",
    "\t\tyield x\n",
    "\n",
    "\tif endpoint and prev != end:\n",
    "\t\tyield end\n",
    "\n",
    "\n",
    "def geomrange(start, end, factor, endpoint=False):\n",
    "\treturn np.fromiter(_geomrange(start, end, factor, endpoint), dtype=int)\n",
    "\n",
    "\n",
    "def geomspace(start, end, num=50, endpoint=True):\n",
    "\tfactor = 1 + ((np.log10(end) - np.log10(start)) / num)\n",
    "\treturn geomrange(start, end, factor, endpoint=endpoint)\n",
    "\n",
    "\n",
    "def _contact_areas(distbins, scaffold_length):\n",
    "\tdistbins = distbins.astype(float)\n",
    "\tscaffold_length = float(scaffold_length)\n",
    "\touter_areas = np.maximum(scaffold_length - distbins[:-1], 0) ** 2\n",
    "\tinner_areas = np.maximum(scaffold_length - distbins[1: ], 0) ** 2\n",
    "\treturn 0.5 * (outer_areas - inner_areas)\n",
    "\t\n",
    "\n",
    "def contact_areas(distbins, region1, region2):\n",
    "\tif region1 == region2:\n",
    "\t\tstart, end = region1\n",
    "\t\tareas = _contact_areas(distbins, end - start)\n",
    "\telse:\n",
    "\t\tstart1, end1 = region1\n",
    "\t\tstart2, end2 = region2\n",
    "\t\tif start2 <= start1:\n",
    "\t\t\tstart1, start2 = start2, start1\n",
    "\t\t\tend1, end2 = end2, end1\n",
    "\t\tareas = (\n",
    "\t\t\t_contact_areas(distbins, end2 - start1) -\n",
    "\t\t\t_contact_areas(distbins, start2 - start1) -\n",
    "\t\t\t_contact_areas(distbins, end2 - end1)\n",
    "\t\t)\n",
    "\t\tif end1 < start2:\n",
    "\t\t\tareas += _contact_areas(distbins, start2 - end1)\n",
    "\t\n",
    "\treturn areas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c91cea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scaling_on_go(input_df):\n",
    "    chromosomes=list(chr_sizes.index[:-3])\n",
    "    \n",
    "    chrom_out = {}\n",
    "    chrom_obs = {}\n",
    "\n",
    "    for chrom in chromosomes:\n",
    "    \n",
    "        start = 0\n",
    "        end = chr_sizes.loc[chrom].values[0]\n",
    "\n",
    "        region1 = (start, end)\n",
    "        region2 = region1\n",
    "\n",
    "        df1 = input_df[input_df[\"chrom1\"] == chrom]\n",
    "\n",
    "        df1 = df1[ \n",
    "            (df1['pos1'] >= region1[0]) & \n",
    "            (df1['pos1'] < region1[1]) & \n",
    "            (df1['pos2'] >= region2[0]) & \n",
    "            (df1['pos2'] < region2[1])\n",
    "        ]\n",
    "        dists = abs(df1['pos2'] - df1['pos1']).values\n",
    "\n",
    "        obs, _ = np.histogram(\n",
    "            dists[(dists >= 1000) & (dists < 100000000)],\n",
    "            bins=distbins)\n",
    "\n",
    "        area = contact_areas(distbins, region1, region2)\n",
    "\n",
    "        chrom_out[chrom] = obs/area\n",
    "        chrom_obs[chrom] = obs\n",
    "    chrom_out[\"Average\"] = np.nanmean(list(chrom_out.values()), axis=0)\n",
    "    chrom_out[\"Distbins\"] = distbins\n",
    "\n",
    "\n",
    "        \n",
    "    return chrom_out, chrom_obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6523c93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in walks, convert to pairwise interactions, convert to cooler\n",
    "\n",
    "#Pair format: \n",
    "#index\tname\tdescription\n",
    "#1\tread_id\tthe ID of the read as defined in fastq files\n",
    "#2\tchrom1\tthe chromosome of the alignment on side 1\n",
    "#3\tpos1\tthe 1-based genomic position of the outer-most (5’) mapped bp on side 1 - I'm using midpoint of mapped fragment\n",
    "#4\tchrom2\tthe chromosome of the alignment on side 2\n",
    "#5\tpos2\tthe 1-based genomic position of the outer-most (5’) mapped bp on side 2 - I'm using midpoint of mapped fragment\n",
    "#6\tstrand1\tthe strand of the alignment on side 1\n",
    "#7\tstrand2\tthe strand of the alignment on side 2\n",
    "\n",
    "#Using the midpoints of the mapped fragments since these are not sonicated, and are long reads.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86f4799e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_full = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    reads =  real_walks_with_summary_firstx[cond]['length_6'][['Query_Name', 'chrom', 'mid', 'Strand']]\n",
    "    reads.columns = ['read_id', 'chrom1', 'pos1', 'strand1']\n",
    "    reads2 = reads.shift()\n",
    "    reads2.columns = ['read_id2', 'chrom2', 'pos2', 'strand2']\n",
    "    pairs = reads.join(reads2)\n",
    "    pairs = pairs[pairs['read_id'] == pairs['read_id2']].reset_index(drop = True)[['read_id', 'chrom1', 'pos1', 'chrom2', 'pos2', 'strand1', 'strand2']]\n",
    "    pairs = pairs.convert_dtypes(infer_objects=True, convert_string=True, convert_integer=True, convert_boolean=True, convert_floating=True)\n",
    "    pairs_full[cond] = pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12c5995",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_full[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4493b1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_sizes = pd.read_table(\"../data/hg38.reduced.chrom.sizes\", header=None, index_col=0)\n",
    "distbins = geomspace(1000, 100000000, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f10442",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_sizes.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "225b6c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove chromosomes not in hg38 reduced, save .pairs file\n",
    "for cond in conditions:\n",
    "    pairs_full[cond] = pairs_full[cond][(pairs_full[cond]['chrom1'].isin(chr_sizes.index)) &\n",
    "                                        (pairs_full[cond]['chrom2'].isin(chr_sizes.index))\n",
    "                                       ].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7ece2ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_full[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67d2f66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/31254050/adding-a-comment-with-to-csv-using-pandas\n",
    "\n",
    "for cond in conditions:\n",
    "    # Open a file in write mode to add the comment\n",
    "    # Then close the file and reopen it with pandas in append mode\n",
    "    with open(f'{outDataDir}/data/220518_{cond}_AllDirect.pairs', 'w') as f:\n",
    "        f.write('#header\\n')\n",
    "        \n",
    "    pairs_full[cond].to_csv(f'{outDataDir}/data/220518_{cond}_AllDirect.pairs', sep = '\\t', \n",
    "                            index = False, header = None, mode = 'a')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d4271e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#then flip pairs to make upper triangular matrix\n",
    "chromsizes_file = \"../data/hg38.reduced.chrom.sizes\"\n",
    "for cond in conditions:\n",
    "    pairsfile = f'{outDataDir}/data/220518_{cond}_AllDirect.pairs'\n",
    "    outputfile = f'{outDataDir}/data/220518_{cond}_AllDirect_Flipped.pairs'\n",
    "    !pairtools flip -c $chromsizes_file -o $outputfile $pairsfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f664bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#then sort pairs to make upper triangular matrix\n",
    "for cond in conditions:\n",
    "    pairsfile = f'{outDataDir}/data/220518_{cond}_AllDirect_Flipped.pairs'\n",
    "    outputfile = f'{outDataDir}/data/220518_{cond}_AllDirect_Sorted.pairs'\n",
    "    !pairtools sort -o $outputfile $pairsfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d569e59",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_full_sorted = {}\n",
    "for cond in conditions:\n",
    "    pairs_full_sorted[cond] = pd.read_csv(f'{outDataDir}/data/220518_{cond}_AllDirect_Sorted.pairs', \n",
    "                                          sep = '\\t', header = None, comment = '#')\n",
    "    pairs_full_sorted[cond].columns = ['read_id', 'chrom1', 'pos1', 'chrom2', 'pos2', 'strand1', 'strand2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56f730f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_full_sorted[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e401fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "output = {}\n",
    "output_obs = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    output[cond], output_obs[cond] = scaling_on_go(pairs_full_sorted[cond])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54de1f34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#normalized by auc\n",
    "#scaling plot\n",
    "fig = plt.figure(figsize=(5, 4))\n",
    "\n",
    "for cond in conditions:\n",
    "    dbins = np.array(output[cond]['Distbins'], dtype = float)\n",
    "    x = np.sqrt(dbins[:-1] * dbins[1:])\n",
    "    y = output[cond]['Average']\n",
    "    yfill = np.nan_to_num(y)\n",
    "    auc = np.trapz(x, yfill)\n",
    "    plt.plot(x[0:], y/-auc, label = cond, lw = 1, color = sampleColors[cond], ls = sampleLineStyles[cond])\n",
    "\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "plt.ylabel(\"P(s)\")\n",
    "plt.xlabel(\"separation (bp)\")\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.1, prop = {'size':8})\n",
    "plt.title('Scaling, normalized to auc')\n",
    "\n",
    "#fixing ticks\n",
    "locmajx = mpl.ticker.LogLocator(base=10,numticks=6) \n",
    "ax.xaxis.set_major_locator(locmajx)\n",
    "locmajy = mpl.ticker.LogLocator(base=10,numticks=6) \n",
    "ax.xaxis.set_major_locator(locmajy)\n",
    "\n",
    "locmin = mpl.ticker.LogLocator(base=10.0,subs=(0.2,0.4,0.6,0.8),numticks=12)\n",
    "ax.xaxis.set_minor_locator(locmin)\n",
    "ax.xaxis.set_minor_formatter(mpl.ticker.NullFormatter())\n",
    "\n",
    "locminy = mpl.ticker.LogLocator(base=10.0,subs=(0.2,0.4,0.6,0.8),numticks=12)\n",
    "ax.yaxis.set_minor_locator(locminy)\n",
    "ax.yaxis.set_minor_formatter(mpl.ticker.NullFormatter())\n",
    "    \n",
    "plt.xlim(1e3, 3e8)\n",
    "plt.tight_layout()\n",
    "\n",
    "fname = f'{outDataDir}/figures/220518_R1R2R3_MR_MC3C_ScalingPlot_AllReads_Direct_Normalized.png'\n",
    "\n",
    "plt.savefig(fname, dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd155dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#derivative plot\n",
    "\n",
    "fig = plt.figure(figsize=(5, 4))\n",
    "\n",
    "for cond in conditions:\n",
    "    dbins = output[cond]['Distbins']\n",
    "    x = np.sqrt(dbins[:-1] * dbins[1:])\n",
    "    y = output[cond]['Average']\n",
    "    dydx = diff(np.log(y))/diff(np.log(x))\n",
    "    dx = np.sqrt(x[:-1] * x[1:])\n",
    "    plt.plot(dx, scipy.ndimage.filters.gaussian_filter1d(dydx, 1.2), label = cond, lw = 1, color = sampleColors[cond],\n",
    "            ls = sampleLineStyles[cond])\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.set_xscale('log')\n",
    "plt.ylabel(\"P(s) Slope\")\n",
    "plt.xlabel(\"separation (bp)\")\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.1, prop = {'size':8})\n",
    "plt.title('Scaling Slope, MC3C')\n",
    "\n",
    "#fixing ticks\n",
    "locmajx = mpl.ticker.LogLocator(base=10,numticks=6) \n",
    "ax.xaxis.set_major_locator(locmajx)\n",
    "\n",
    "locmin = mpl.ticker.LogLocator(base=10.0,subs=(0.2,0.4,0.6,0.8),numticks=12)\n",
    "ax.xaxis.set_minor_locator(locmin)\n",
    "ax.xaxis.set_minor_formatter(mpl.ticker.NullFormatter())\n",
    "\n",
    "plt.xlim(1e3, 3e8)\n",
    "plt.tight_layout()\n",
    "\n",
    "fname = f'{outDataDir}/figures/220518_R1R2R3_MR_ScalingPlotSlope_AllReads_Direct.png'\n",
    "\n",
    "plt.savefig(fname, dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513ec9c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Alignment length distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5291fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "outDataDir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f5780d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in alignments for each condition\n",
    "alignmentDir = #directory_of_raw_alignments\n",
    "\n",
    "aligned_reads = pd.DataFrame()\n",
    "for cond in conditions:\n",
    "    df = pd.read_csv(f'{alignmentDir}/{long_names[cond]}.hg38.minimap2.output.paf', header = None,\n",
    "                                  sep = '\\t')\n",
    "    df.columns = ['Query_Name', 'Query_Length', 'Query_Start', 'Query_End',\n",
    "                                  'Strand', 'chrom', 'Target_ChrSize', 'start', \n",
    "                                  'end', 'Match_Length', 'Alignment_Length', 'Mapping_Quality', \n",
    "                  'TypeOfAln', 'NumberOfMinimizers', 'ChainingScorePrimary', \n",
    "                  'ChainingScoreSecondary', 'ApproxPerBaseDivergence', 'RepetitiveSeedLength'] \n",
    "    df['Condition'] = cond\n",
    "    \n",
    "    aligned_reads = aligned_reads.append(df).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38edc0ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_reads.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f292515b",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignednum = aligned_reads.pivot_table(index = 'Condition', values = 'Query_Name', aggfunc = pd.Series.nunique)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a23043c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignednum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db708dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragmentnum = pd.DataFrame(aligned_reads.groupby(['Condition', 'Query_Name']).size()).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c306c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragmentnum.columns = ['Condition', 'Query_Name', 'Count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a0b5ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "repdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3302771b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragmentnum['Label'] = np.nan\n",
    "fragmentnum['Replicate'] = np.nan\n",
    "\n",
    "for cond in conditions:\n",
    "    fragmentnum['Label'][fragmentnum['Condition'] == cond] = labeldict[cond]\n",
    "    fragmentnum['Replicate'][fragmentnum['Condition'] == cond] = repdict[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aa064ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragmentnum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43ec2b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    g = sns.displot(kind='hist', data=fragmentnum, col='Replicate', x='Count', hue='Label', \n",
    "                    palette = cmap_bar, log_scale = False, common_norm = False, common_bins = True, \n",
    "                    lw = 2, facet_kws={'sharey': True}, height = 3, element = 'step', fill = False,\n",
    "                    stat = 'density', binwidth = 1, hue_order = ['t0 Prometa', 't2 Ana/Telo', 't4 G1 DMSO', 't4 G1 ICRF-193', 't8 G1 DMSO', 't8 G1 ICRF-193'])\n",
    "    g._legend.set_title('Sample')\n",
    "    g.set_xlabels('Number Of Fragments')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_R1R2R3_Separate_Density_FragmentNumber_Kde_DropNA.png', bbox_inches = 'tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe0bd51",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read length (query_length) of aligned reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "275e7647",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15ac4ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_length = pd.DataFrame(aligned_reads.groupby(['Condition', 'Query_Name']).mean('Query_Length')).reset_index()[['Condition', 'Query_Name', 'Query_Length']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a09b7682",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9305937",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_length['Label'] = np.nan\n",
    "read_length['Replicate'] = np.nan\n",
    "\n",
    "for cond in conditions:\n",
    "    read_length['Label'][read_length['Condition'] == cond] = labeldict[cond]\n",
    "    read_length['Replicate'][read_length['Condition'] == cond] = repdict[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4cc546a",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9388ab40",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    g = sns.displot(kind='kde', data=read_length, col='Replicate', x='Query_Length', hue='Label', \n",
    "                    palette = cmap_bar, log_scale = True, common_norm = False, common_grid = True, \n",
    "                    lw = 2, facet_kws={'sharey': True}, height = 3, hue_order = ['t0 Prometa', 't2 Ana/Telo', 't4 G1 DMSO', 't4 G1 ICRF-193', 't8 G1 DMSO', 't8 G1 ICRF-193'])\n",
    "    g._legend.set_title('Sample')\n",
    "    g.set_xlabels('Read Length')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_R1R3R4_Separate_Density_ReadLength_LogScaleKde.png', bbox_inches = 'tight', dpi = 300))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fb4ea8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run percent cis on pairs files\n",
    "cis_percent = {}\n",
    "for cond in conditions:\n",
    "    pairs_full_sorted = pd.read_csv(f'{outDataDir}/data/220518_{cond}_AllDirect_Sorted.pairs', \n",
    "                                          sep = '\\t', header = None, comment = '#')\n",
    "    pairs_full_sorted.columns = ['read_id', 'chrom1', 'pos1', 'chrom2', 'pos2', 'strand1', 'strand2']\n",
    "    cis_count = len(pairs_full_sorted[pairs_full_sorted['chrom1'] == pairs_full_sorted['chrom2']])\n",
    "    total_count = len(pairs_full_sorted)\n",
    "    cis_percent[cond] = (cis_count/total_count) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b609bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_percent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca7a24d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Intermingling plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af4d90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For each of the above, can separate by A vs B compartments, or by number of chromosomes visited\n",
    "comp_types = ['A', 'AB', 'B']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551cc759",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load intermingling results as pickle file for later use since takes a long time to run\n",
    "f = open(f'{outDataDir}/data/220518_MRICRF_R1R2R3_InterminglingSlidingWindowSweep.pkl', 'rb')\n",
    "Intermingling_Sliding_Window_Sweep = pickle.load(f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d055b971",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_Sliding_Window_Sweep_CT_Counts = {}\n",
    "\n",
    "for window_size in [4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]:\n",
    "    Intermingling_Sliding_Window_Sweep_CT_Counts[f'{window_size}'] = pd.crosstab(\n",
    "        index=[Intermingling_Sliding_Window_Sweep[f'{window_size}']['Condition'],\n",
    "               Intermingling_Sliding_Window_Sweep[f'{window_size}']['Window_Midpoint'],\n",
    "               Intermingling_Sliding_Window_Sweep[f'{window_size}']['Label'],\n",
    "               Intermingling_Sliding_Window_Sweep[f'{window_size}']['Replicate']\n",
    "              ],\n",
    "        columns=Intermingling_Sliding_Window_Sweep[f'{window_size}']['Walk_Comp_Type'], \n",
    "        normalize = False).reset_index().melt(id_vars = [\n",
    "        'Condition', \n",
    "        'Window_Midpoint',\n",
    "        'Label',\n",
    "        'Replicate'\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f5e35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(24, 12))\n",
    "gs = GridSpec(nrows= 2, ncols=4, wspace = 1, hspace = .4)\n",
    "cmap_bar = sns.color_palette(['#878787','#E1B7A3','#17BECF', '#D62728', '#0D6871', '#751616'])\n",
    "\n",
    "for i, window_size in enumerate([4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]):\n",
    "    ax = plt.subplot(gs[i])\n",
    "    sns.lineplot(data = Intermingling_Sliding_Window_Sweep_CT_Counts[f'{window_size}'][\n",
    "        Intermingling_Sliding_Window_Sweep_CT_Counts[f'{window_size}']['Walk_Comp_Type'] == 'All'], \n",
    "                 x = 'Window_Midpoint', y = 'value', hue = 'Label', style = 'Replicate', palette = cmap_bar)\n",
    "    plt.ylabel('Number of Walks')\n",
    "    plt.xlabel('Window Midpoint (bp)')\n",
    "    plt.title(f'Walk Number Per {window_size} bp Window')\n",
    "plt.savefig(f'{outDataDir}/figures/220518_MRICRF_AllSamples_WalkNumberPerWindow_AllComps_Separate_WindowSizeSweep_GoodChroms.png', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed5a28b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Heatmaps of each comparison of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae4a5f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for heatmaps - combine all replicates\n",
    "Intermingling_Sliding_Window_Sweep_CT = {}\n",
    "\n",
    "for window_size in [4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]:\n",
    "    Intermingling_Sliding_Window_Sweep_CT[f'{window_size}'] = pd.crosstab(\n",
    "        index=[Intermingling_Sliding_Window_Sweep[f'{window_size}']['Label'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Window_Midpoint'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Walk_Comp_Type'],\n",
    "          ],\n",
    "    columns=Intermingling_Sliding_Window_Sweep[f'{window_size}']['Inter_Largest_Step_Side_Step_Num'], \n",
    "    normalize = False).reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0af3af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_SlidingWindow_Melt_Sweep = {}\n",
    "\n",
    "for window_size in [4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]:\n",
    "    Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}'] = Intermingling_Sliding_Window_Sweep_CT[f'{window_size}'].melt(id_vars = [\n",
    "    'Label', \n",
    "    'Window_Midpoint',\n",
    "    'Walk_Comp_Type',\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18eed751",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_SlidingWindow_Sweep_ToPlot = {}\n",
    "for window_size in [4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]:\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}']= Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}'].loc[\n",
    "        (Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}']['Inter_Largest_Step_Side_Step_Num'] == 2)\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd7e8be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = ['t0 Prometa', 't2 Ana/Telo', 't4 G1 DMSO', 't4 G1 ICRF-193', 't8 G1 DMSO', 't8 G1 ICRF-193']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d75ee7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_df = Intermingling_SlidingWindow_Sweep_ToPlot[f'{1.6e7}'].merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{1.4e7}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num'], suffixes = (f'_{1.6e7}', f'_{1.4e7}')).merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{1.2e7}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num']).merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{1e7}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num'], suffixes = (f'_{1.2e7}', f'_{1e7}')).merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{8e6}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num']).merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{6e6}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num'], suffixes = (f'_{8e6}', f'_{6e6}')).merge(\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{4e6}'], how = 'outer', \n",
    "    on = ['Label', 'Window_Midpoint', 'Walk_Comp_Type', \n",
    "          'Inter_Largest_Step_Side_Step_Num']).sort_values(by = ['Label', 'Window_Midpoint', 'Walk_Comp_Type'])\n",
    "\n",
    "heatmap_dfs_bylabel = {}\n",
    "\n",
    "for cond in labels:\n",
    "    df = heatmap_df[heatmap_df['Label'] == cond]\n",
    "    \n",
    "    df = df[(df['Walk_Comp_Type'] == 'All')][[\n",
    "        'Window_Midpoint', 'value_16000000.0', 'value_14000000.0', 'value_12000000.0',\n",
    "        'value_10000000.0', 'value_8000000.0','value_6000000.0', 'value']].set_index('Window_Midpoint', drop = True)\n",
    "\n",
    "    df.columns = [16, 14, 12, 10, 8, 6, 4]\n",
    "    \n",
    "    heatmap_dfs_bylabel[cond] = df.transpose()\n",
    "    heatmap_dfs_bylabel[cond].columns = list(range(2, 28))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72395428",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_dfs_bylabel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31fa69fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cooltools.lib.plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02a334a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(24, 9))\n",
    "\n",
    "gs = GridSpec(nrows= 3, ncols=4, wspace = 0.3, hspace = 1)\n",
    "for i, cond in enumerate(labels):\n",
    "    ax = plt.subplot(gs[i])\n",
    "    sns.heatmap(heatmap_dfs_bylabel[cond], square = False, \n",
    "                vmin=0.4, \n",
    "                vmax=0.65, \n",
    "                cmap = 'fall', cbar_kws={'label': 'IM'},\n",
    "                mask=heatmap_dfs_bylabel[cond].isnull())\n",
    "    plt.ylabel('Window Size (Mb)')\n",
    "    plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "    plt.title(cond)\n",
    "    #ax.set_aspect('equal')\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/220518_Heatmaps_SlidingWindowSweep_MRICRF_GoodChroms_4to16Mb.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee467f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#log2 ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "923ea052",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl_labels = ['t2 Ana/Telo', 't4 G1 DMSO', 't8 G1 DMSO', 't4 G1 DMSO', 't8 G1 DMSO', 't8 G1 DMSO', 't2 Ana/Telo', 't4 G1 DMSO', 't2 Ana/Telo', 't8 G1 DMSO']\n",
    "treat_labels = ['t0 Prometa', 't0 Prometa', 't0 Prometa', 't2 Ana/Telo', 't2 Ana/Telo', 't4 G1 DMSO', 't4 G1 ICRF-193', 't4 G1 ICRF-193', 't8 G1 ICRF-193', 't8 G1 ICRF-193']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa64903",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(24, 9))\n",
    "\n",
    "gs = GridSpec(nrows= 3, ncols=4, wspace = 0.3, hspace = 1)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = plt.subplot(gs[i])\n",
    "    log2df = np.log2(heatmap_dfs_bylabel[treatcond]) - np.log2(heatmap_dfs_bylabel[ctrlcond])                                      \n",
    "    sns.heatmap(log2df, square = False, vmin=-.5, vmax=.5, cmap = 'coolwarm', cbar_kws={'label': 'log2FC treat/ctrl'})\n",
    "    plt.ylabel('Window Size (Mb)')\n",
    "    plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "    plt.title(f'{treatcond} vs {ctrlcond}')\n",
    "    #ax.set_aspect('equal')\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/220518_InterminglingWindowSizeSweep_Heatmaps_TreatvsCtrl_goodchroms_CellCycleDMSO_4to16Mb.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b4caad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Just 4mb window size - all samples together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793d4044",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_4mb_only = pd.DataFrame()\n",
    "\n",
    "for cond in labels:\n",
    "    df_4mb_only = df_4mb_only.append(heatmap_dfs_bylabel[cond].loc[4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9b3d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_4mb_only.index = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61329b3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_4mb_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1320a719",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(\n",
    "    df_4mb_only, \n",
    "    square = False,       \n",
    "    vmin=0.35,                 \n",
    "    vmax=0.65,                 \n",
    "    cmap = 'fall', \n",
    "    cbar_kws={'label': 'IM'},                \n",
    "    mask=df_4mb_only.isnull())\n",
    "plt.ylabel('Sample')\n",
    "plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "plt.savefig(f'{outDataDir}/figures/220518_MRICRF_IMHeatmap_4MbWindow.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb424ffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_12mb_only = pd.DataFrame()\n",
    "\n",
    "for cond in labels:\n",
    "    df_12mb_only = df_12mb_only.append(heatmap_dfs_bylabel[cond].loc[12])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077a2375",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_12mb_only.index = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b89b84e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_12mb_only.dropna(axis = 'columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bacded5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(\n",
    "    df_12mb_only.dropna(axis = 'columns'), \n",
    "    square = False,       \n",
    "    vmin=0.4,                 \n",
    "    vmax=0.65,                 \n",
    "    cmap = 'fall', \n",
    "    cbar_kws={'label': 'IM'},                \n",
    "    #mask=df_12mb_only.isnull()\n",
    ")\n",
    "plt.ylabel('Sample')\n",
    "plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "plt.savefig(f'{outDataDir}/figures/220518_MRICRF_IMHeatmap_12MbWindow.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b944edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#line plots with error and permutations\n",
    "\n",
    "window_sizes = [4e6, 6e6, 8e6, 1e7, 1.2e7, 1.4e7, 1.6e7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52dea414",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_Sliding_Window_Sweep_CT = {}\n",
    "\n",
    "for window_size in window_sizes:\n",
    "    Intermingling_Sliding_Window_Sweep_CT[f'{window_size}'] = pd.crosstab(\n",
    "        index=[Intermingling_Sliding_Window_Sweep[f'{window_size}']['Condition'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Window_Midpoint'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Label'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Walk_Comp_Type'],\n",
    "           Intermingling_Sliding_Window_Sweep[f'{window_size}']['Replicate']\n",
    "          ],\n",
    "    columns=Intermingling_Sliding_Window_Sweep[f'{window_size}']['Inter_Largest_Step_Side_Step_Num'], \n",
    "    normalize = 'index').reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "733cf68b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_SlidingWindow_Melt_Sweep = {}\n",
    "\n",
    "for window_size in window_sizes:\n",
    "    Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}'] = Intermingling_Sliding_Window_Sweep_CT[f'{window_size}'].melt(id_vars = [\n",
    "    'Condition', \n",
    "    'Window_Midpoint',\n",
    "    'Label',\n",
    "    'Walk_Comp_Type',\n",
    "    'Replicate'\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "080e30d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_SlidingWindow_Sweep_ToPlot = {}\n",
    "for window_size in window_sizes:\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}']= Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}'].loc[\n",
    "        (Intermingling_SlidingWindow_Melt_Sweep[f'{window_size}']['Inter_Largest_Step_Side_Step_Num'] == 2)\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "848925c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#permutations\n",
    "iterations = range(100)\n",
    "\n",
    "#read in pickled intermingling data for each, extract data to plot\n",
    "Permuted_Intermingling_SlidingWindow_Sweep_ToPlot = {}\n",
    "\n",
    "for window_size in window_sizes:\n",
    "    Permuted_Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}'] = pd.DataFrame()\n",
    "\n",
    "for i in iterations:\n",
    "    for window_size in window_sizes:\n",
    "        #run crosstab for each iteration, each window size, melt, and combine into full dataframe\n",
    "        #read in pickled intermingling data\n",
    "        f = open(f'{outDataDir}/data/permutations/220518_MRICRF_R1R2R3_PermutedInterminglingSlidingWindowSweep_window{window_size}.iter{i}.pkl', 'rb')\n",
    "        Permuted_Intermingling_Sliding_Window_Sweep_OneSize = pickle.load(f)\n",
    "        f.close()\n",
    "        Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Iteration'] = i\n",
    "        \n",
    "        Permuted_CT = pd.crosstab(\n",
    "            index=[\n",
    "                Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Condition'],\n",
    "                Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Window_Midpoint'],\n",
    "                Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Label'],\n",
    "                Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Iteration'],\n",
    "                Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Walk_Comp_Type']\n",
    "              ],\n",
    "            columns=Permuted_Intermingling_Sliding_Window_Sweep_OneSize['Inter_Largest_Step_Side_Step_Num'], \n",
    "            normalize = 'index').reset_index()\n",
    "\n",
    "        Permuted_Melt = Permuted_CT.melt(id_vars = [\n",
    "            'Condition', \n",
    "            'Window_Midpoint',\n",
    "            'Label',\n",
    "            'Iteration',\n",
    "            'Walk_Comp_Type'\n",
    "        ])\n",
    "\n",
    "        Permuted_ToPlot = Permuted_Melt.loc[\n",
    "            (Permuted_Melt['Inter_Largest_Step_Side_Step_Num'] == 2)]\n",
    "\n",
    "        Permuted_Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}'] = Permuted_Intermingling_SlidingWindow_Sweep_ToPlot[\n",
    "            f'{window_size}'].append(Permuted_ToPlot, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd450752",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_BothWalkTypes_ToPlot = {}\n",
    "\n",
    "for window_size in window_sizes:\n",
    "    Permuted_Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}']['Walk_Type'] = 'Permuted'\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}']['Walk_Type'] = 'Real'\n",
    "    Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}']['Iteration'] = np.nan\n",
    "    \n",
    "    Intermingling_BothWalkTypes_ToPlot[f'{window_size}'] = Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}'].append(Permuted_Intermingling_SlidingWindow_Sweep_ToPlot[f'{window_size}'], ignore_index = True)\n",
    "\n",
    "    Intermingling_BothWalkTypes_ToPlot[f'{window_size}'].columns = ['Condition', 'Window Midpoint', 'Sample', 'Walk_Comp_Type', 'Replicate',\n",
    "       'Inter_Largest_Step_Side_Step_Num', 'Fraction Intermingled', 'Walk Type', 'Iteration']\n",
    "    \n",
    "    Intermingling_BothWalkTypes_ToPlot[f'{window_size}']['Window Size'] = window_size\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96078081",
   "metadata": {},
   "outputs": [],
   "source": [
    "Intermingling_BothWalkTypes_ToPlot_AllWindowSizes = pd.DataFrame()\n",
    "for window_size in window_sizes:\n",
    "    Intermingling_BothWalkTypes_ToPlot_AllWindowSizes = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes.append(Intermingling_BothWalkTypes_ToPlot[f'{window_size}'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff330a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = ['t0 Prometa', 't2 Ana/Telo', 't4 G1 DMSO', 't4 G1 ICRF-193', 't8 G1 DMSO', 't8 G1 ICRF-193']\n",
    "ctrl_labels = ['t0 Prometa', 't0 Prometa', 't0 Prometa', 't2 Ana/Telo', 't2 Ana/Telo', 't4 G1 DMSO', 't2 Ana/Telo', 't4 G1 DMSO', 't2 Ana/Telo', 't8 G1 DMSO']\n",
    "treat_labels = ['t2 Ana/Telo', 't4 G1 DMSO', 't8 G1 DMSO', 't4 G1 DMSO', 't8 G1 DMSO', 't8 G1 DMSO', 't4 G1 ICRF-193', 't4 G1 ICRF-193', 't8 G1 ICRF-193', 't8 G1 ICRF-193']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe90c797",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(20, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'All'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.2, 0.85)\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_FractionIntermingled_RealvsPermuted_AllComps_{i}_AllWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0de46ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(20, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'All'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, ci = 'sd')\n",
    "    plt.ylim(0.2, 0.85)\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_FractionIntermingled_RealvsPermuted_AllComps_{i}_AllWindows_sd.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87e94708",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4Mb window only\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 4e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'All'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.1, 0.8)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_FractionIntermingled_RealvsPermuted_AllComps_{i}_4MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db75f6b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12Mb window only\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 12e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'All'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.35, 0.7)\n",
    "    plt.xlim(0.5e7, 2.4e7)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/220518_MRICRF_FractionIntermingled_RealvsPermuted_AllComps_{i}_12MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "532f0040",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_dfs_bylabel_bycomp = {}\n",
    "\n",
    "for cond in labels:\n",
    "    heatmap_dfs_bylabel_bycomp[cond] = {}\n",
    "    for comp in comp_types:\n",
    "        df = heatmap_df[heatmap_df['Label'] == cond]\n",
    "\n",
    "        df = df[(df['Walk_Comp_Type'] == comp)][[\n",
    "            'Window_Midpoint', 'value_16000000.0', 'value_14000000.0', 'value_12000000.0',\n",
    "            'value_10000000.0', 'value_8000000.0','value_6000000.0', 'value']].set_index('Window_Midpoint', drop = True)\n",
    "\n",
    "        df.columns = [16, 14, 12, 10, 8, 6, 4]\n",
    "\n",
    "        heatmap_dfs_bylabel_bycomp[cond][comp] = df.transpose()\n",
    "        heatmap_dfs_bylabel_bycomp[cond][comp].columns = list(range(2, 48))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e35c2b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_dfs_bylabel_bycomp[cond][comp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b3b45d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(24, 18))\n",
    "\n",
    "gs = GridSpec(nrows= 6, ncols=4, wspace = 0.3, hspace = 1)\n",
    "\n",
    "for i, cond in enumerate(labels):\n",
    "    for j, comp in enumerate(comp_types):\n",
    "        ax = plt.subplot(gs[i, j])\n",
    "        sns.heatmap(heatmap_dfs_bylabel_bycomp[cond][comp], square = False, \n",
    "                vmin=0.4, \n",
    "                vmax=0.7, \n",
    "                cmap = 'fall', cbar_kws={'label': 'IM'},\n",
    "                mask=heatmap_dfs_bylabel_bycomp[cond][comp].isnull())\n",
    "        plt.ylabel('Window Size (Mb)')\n",
    "        plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "        plt.title(f'{cond}, {comp}')\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/Heatmaps_SlidingWindowSweep_AllChroms_ByComp.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70f5e938",
   "metadata": {},
   "outputs": [],
   "source": [
    "#log2 ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3172f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(24, 18))\n",
    "\n",
    "gs = GridSpec(nrows= 6, ncols=4, wspace = 0.3, hspace = 1)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    for j, comp in enumerate(comp_types):\n",
    "        ax = plt.subplot(gs[i, j])\n",
    "        log2df = np.log2(heatmap_dfs_bylabel_bycomp[treatcond][comp]) - np.log2(heatmap_dfs_bylabel_bycomp[ctrlcond][comp])                                      \n",
    "        sns.heatmap(log2df, square = False, vmin=-.5, vmax=.5, cmap = 'coolwarm', cbar_kws={'label': 'log2FC treat/ctrl'})\n",
    "        plt.ylabel('Window Size (Mb)')\n",
    "        plt.xlabel('Genomic Distance Between Interacting Domains \\n(Window Midpoint, Mb)')\n",
    "        plt.title(f'{treatcond} vs {ctrlcond}, {comp}')\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/InterminglingWindowSizeSweep_Heatmaps_TreatvsCtrl_ByComp.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f68e53f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12Mb window only, B comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 12e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'B'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.35, 0.9)\n",
    "    plt.xlim(0.5e7, 4.4e7)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/Rad21ICRF_FractionIntermingled_RealvsPermuted_BComp_{i}_12MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050bcf58",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12Mb window only, A comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 12e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'A'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.35, 0.9)\n",
    "    plt.xlim(0.5e7, 4.4e7)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/Rad21ICRF_FractionIntermingled_RealvsPermuted_AComp_{i}_12MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a559762e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12Mb window only, AB comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 12e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'AB'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.35, 0.9)\n",
    "    plt.xlim(0.5e7, 4.4e7)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/Rad21ICRF_FractionIntermingled_RealvsPermuted_ABComp_{i}_12MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d694569",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4Mb window only, all comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 4e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'All'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.1, 1)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/Rad21ICRF_FractionIntermingled_RealvsPermuted_AllComp_{i}_4MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e96057",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4Mb window only, B comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 4e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'B'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.1, 1)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/FractionIntermingled_RealvsPermuted_BComp_{i}_4MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40d427b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4Mb window only, A comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 4e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'A'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.1, 1)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/FractionIntermingled_RealvsPermuted_AComp_{i}_4MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "564c6aaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4Mb window only, AB comp types\n",
    "\n",
    "fig = plt.figure(figsize=(3, 2))\n",
    "gs = GridSpec(nrows= 10, ncols=1, wspace = 1, hspace = .4)\n",
    "\n",
    "for i, (ctrlcond, treatcond) in enumerate(zip(ctrl_labels, treat_labels)):\n",
    "    ax = gs[i]\n",
    "    df = Intermingling_BothWalkTypes_ToPlot_AllWindowSizes[\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Sample'].isin([ctrlcond, treatcond])) &\n",
    "        (Intermingling_BothWalkTypes_ToPlot_AllWindowSizes['Window Size'] == 4e6)\n",
    "    ]\n",
    "    cmap_bar = sns.color_palette([labelPlotColors[df['Sample'].unique()[0]], labelPlotColors[df['Sample'].unique()[1]]])\n",
    "    sns.relplot(data = df[df['Walk_Comp_Type'] == 'AB'],\n",
    "             x = 'Window Midpoint', y = 'Fraction Intermingled', hue = 'Sample', palette = cmap_bar, kind = 'line', \n",
    "                style = 'Walk Type', col = 'Window Size', height = 2, aspect = 1.5, seed = 25)\n",
    "    plt.ylim(0.1, 1)\n",
    "    plt.xlabel('Genomic Distance \\nBetween Interacting Domains \\n (Window Midpoint, bp)')\n",
    "    plt.ylabel('IM')\n",
    "\n",
    "    plt.savefig(f'{outDataDir}/figures/FractionIntermingled_RealvsPermuted_ABComp_{i}_4MbWindows.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
